lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/IMRUDriver.cs

```diff
@@ -98,181 +128,457 @@ internal sealed class IMRUDriver<TMapInput, TMapOutput, TResult, TPartitionType>
-                return;
+                // Change the system state to SubmittingTasks
+                _systemState.MoveNext(SystemStateEvent.AllContextsAreReady);
+                SubmitTasks();
             }
+        }
 
-            Logger.Log(Level.Info,
-                string.Format("Evaluator with Id: {0} failed with Exception: {1}", value.Id, value.EvaluatorException));
-            int currFailedEvaluators = Interlocked.Increment(ref _currentFailedEvaluators);
-            if (currFailedEvaluators > _allowedFailedEvaluators)
+        /// <summary>
+        /// This method is responsible to prepare for the task submission then call SubmitTasks in TaskManager.
+        /// It is called in both first time and recovery scenarios.
+        /// Creates a new Communication Group and adds Group Communication Operators
+        /// For each context, adds a task to the communication group.
+        /// After all the tasks are added to the group, for each task, gets GroupCommTaskConfiguration from IGroupCommDriver 
+        /// and merges it with the task configuration.
+        /// When all the tasks are added, calls TaskManager to SubmitTasks().
+        /// </summary>
+        private void SubmitTasks()
+        {
+            Logger.Log(Level.Info, string.Format("SubmitTasks with system state :" + _systemState.CurrentState));
+            if (!_isFirstTry)
             {
-                Exceptions.Throw(new MaximumNumberOfEvaluatorFailuresExceededException(_allowedFailedEvaluators),
-                    Logger);
+                _taskManager.Reset();
+                _groupCommDriver.RemoveCommunicationGroup(IMRUConstants.CommunicationGroupName);
             }
 
-            _serviceAndContextConfigurationProvider.RecordEvaluatorFailureById(value.Id);
-            bool isMaster = _serviceAndContextConfigurationProvider.IsMasterEvaluatorId(value.Id);
+            var commGroup = AddCommunicationGroupWithOperators();
+            _perMapperConfiguration = ConstructPerMapperConfigStack(_totalMappers);
 
-            // If failed evaluator is master then ask for master 
-            // evaluator else ask for mapper evaluator
-            if (!isMaster)
+            var taskIdAndContextMapping = new Dictionary<string, IActiveContext>();
+            foreach (var activeContext in _contextManager.ActiveContexts)
             {
-                Logger.Log(Level.Info, string.Format("Requesting a replacement map Evaluator for {0}", value.Id));
-                RequestMapEvaluators(1);
+                var taskId = _evaluatorManager.IsMasterEvaluatorId(activeContext.EvaluatorId) ?
+                    _groupCommDriver.MasterTaskId :
+                    GetMapperTaskIdByEvaluatorId(activeContext.EvaluatorId);
+                commGroup.AddTask(taskId);
+                taskIdAndContextMapping.Add(taskId, activeContext);
             }
-            else
+
+            foreach (var mapping in taskIdAndContextMapping)
```

Comment:
yes, same key. 
In the second loop, we call GetGroupCommTaskConfiguration(), that requires all the tasks are added to the group first.